services/oauth/v2/oauth.go

```diff
@@ -0,0 +1,384 @@
+package v2
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	rudderSync "github.com/rudderlabs/rudder-go-kit/sync"
+	obskit "github.com/rudderlabs/rudder-observability-kit/go/labels"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	router_utils "github.com/rudderlabs/rudder-server/router/utils"
+)
+
+func WithCache(cache Cache) func(*OAuthHandler) {
+	return func(h *OAuthHandler) {
+		h.Cache = cache
+	}
+}
+
+func WithLocker(lock *rudderSync.PartitionRWLocker) func(*OAuthHandler) {
+	return func(h *OAuthHandler) {
+		h.CacheMutex = lock
+	}
+}
+
+func WithExpirationTimeDiff(expirationTimeDiff time.Duration) func(*OAuthHandler) {
+	return func(h *OAuthHandler) {
+		h.ExpirationTimeDiff = expirationTimeDiff
+	}
+}
+
+func WithLogger(parentLogger logger.Logger) func(*OAuthHandler) {
+	return func(h *OAuthHandler) {
+		h.Logger = parentLogger
+	}
+}
+
+// NewOAuthHandler returns a new instance of OAuthHandler
+func NewOAuthHandler(provider TokenProvider, options ...func(*OAuthHandler)) *OAuthHandler {
+	cpTimeoutDuration := config.GetDuration("HttpClient.oauth.timeout", 30, time.Second)
```

Comment:
Can we avoid using global's and pass config as a dependency in the constructor?
```suggestion
func NewOAuthHandler(conf *config.Config, provider TokenProvider, options ...func(*OAuthHandler)) *OAuthHandler {
	cpTimeoutDuration := conf.GetDuration("HttpClient.oauth.timeout", 30, time.Second)
```

Reply:
As this NewOAuthHandler is being called in multiple places and we don't have a config available at the caller level created a default config.